platform/zoul/dev/hp206c.c

```diff
@@ -0,0 +1,194 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+/*---------------------------------------------------------------------------*/
+/**
+ * \addtogroup zoul-sensor
+ * @{
+ *
+ * \file
+ *         HP206C Barometer and altimeter sensor driver
+ *
+ * \author
+ *         Javi Sanchez <jsanchez@zolertia.com>
+ */
+/*---------------------------------------------------------------------------*/
+#include <stdio.h>
+#include "contiki.h"
+#include "dev/i2c.h"
+#include "dev/hp206c.h"
+#include "lib/sensors.h"
+/*---------------------------------------------------------------------------*/
+#define DEBUG 0
+#if DEBUG
+#define PRINTF(...) printf(__VA_ARGS__)
+#else
+#define PRINTF(...)
+#endif
+/*---------------------------------------------------------------------------*/
+static uint8_t enabled;
+/*---------------------------------------------------------------------------*/
+static int
+status(int type)
+{
+  switch(type) {
+  case SENSORS_ACTIVE:
+  case SENSORS_READY:
+    return enabled;
+  }
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+static uint16_t
+hp206c_read_reg(uint8_t reg, uint8_t *buf, uint8_t num)
+{
+  if((buf == NULL) || (num <= 0)) {
+    return HP206C_ERROR;
+  }
+
+  i2c_master_enable();
+  if(i2c_single_send(HP206C_ADDR, reg) == I2C_MASTER_ERR_NONE) {
+    if(i2c_burst_receive(HP206C_ADDR, buf, num) == I2C_MASTER_ERR_NONE) {
+      return HP206C_SUCCESS;
+    }
+  }
+  return HP206C_ERROR;
+}
+/*---------------------------------------------------------------------------*/
+static int32_t
+hpc206c_convert(int32_t val)
+{
+  if((val & 0xF00000) == 0xF00000) {
+    val = (val & 0x0FFFFF) ^ 0xFFFFF;
+    return -val;
+  } else {
+    return val;
+  }
+  return HP206C_ERROR;
+}
+/*---------------------------------------------------------------------------*/
+static int
+hp206c_read(uint8_t variable, int32_t *rd)
+{
+  uint8_t buf[3];
+
+  if((variable != HP206C_READ_P) && (variable != HP206C_READ_A)
+     && (variable != HP206C_READ_T) && (variable != HP206C_READ_PT)
+     && (variable != HP206C_READ_AT) && (variable != HP206C_READ_CAL)) {
+    PRINTF("HP206C: invalid sensor requested\n");
+    return HP206C_ERROR;
+  }
+
+  if(hp206c_read_reg(variable, buf, 3) == HP206C_SUCCESS) {
+    *rd = ((buf[0] << 16) | (buf[1] << 8) | buf[2]);
+    //*rd = (uint32_t *)((buf[0] << 16) | (buf[1] << 8) | buf[2]);
+    *rd = hpc206c_convert(*rd);
+    PRINTF("after converter read= %lu\n", *rd);
```

Comment:
I would prefer to have something like `PRINTF(HP206C: xxxx)`, mostly upon debugging to know where the prints are coming from